### ISSUE TITLE
Stream quality shortcuts unavailable?

### ISSUE THREAD

**Issue by Vangelis66** [streamlink/streamlink#89]
Hello 

I am currently testing v0.0.2 of the python script on Windows Vista SP2 x86, 
using the "portable" windows package kindly offered by @RosadinTV 

In livestreamer, I could issue: 

`livestreamer -l debug --http-proxy "proxyhost:80" -o "itv1-1800k.flv" "http://www.itv.com/hub/itv" best` 

and it would work. In fact, issuing: 

`livestreamer -l debug --http-proxy "proxyhost:80" -o "itv1-1800k.flv" "http://www.itv.com/hub/itv"`

would inform me that: 
```
[cli][info] Found matching plugin itvhub for URL http://www.itv.com/hub/itv
Available streams: 1200k (worst), 1500k, 1800k (best)
```
indicating the **worst** and **best** stream quality variants. 
With streamlink, I only get available bitrates: 
```
[cli][info] Found matching plugin itvplayer for URL http://www.itv.com/hub/itv
Available streams: 1200, 1500, 1800
```
so specifying "best" beforehand in my command doesn't work... 

```
[cli][info] Found matching plugin itvplayer for URL http://www.itv.com/hub/itv
error: The specified stream(s) 'best' could not be found.
       Available streams: 1500, 1200, 1800
```

This means I have to run streamlink twice: 
Once to be informed of available bitrates, 
a second time specifying the highest bitrate (indicated by first run):  

`streamlink -l debug --http-proxy "proxyhost:80" -o "itv1-1800k.flv" "http://www.itv.com/hub/itv" 1800`

Am I getting something wrong here?

Thanks for your efforts in resurrecting what appears to be a staled project!

**Comment by RosadinTV** [comment]
I tried and yes, you are right.
For some reason in Python 3.5.2 the stream qualities didnt get the k ending.
I mean with this stream in Python 2.7 returns: Available streams: 1200k (worst), 1500k, 1800k (best) 
In Python 3.5.2 returns: Available streams: 1800, 1500, 1200
Maybe somebody of the staff know a fix for this issue.

**Comment by Vangelis66** [comment]
@RosadinTV

Many thanks for looking into this. 

> with this stream in Python 2.7 returns: 
> Available streams: 1200k (worst), 1500k, 1800k (best) 

Pending an intervention from the main devs, 
could you be so kind as to compile a win32 "portable" package 
(like your second one, where Python and Streamlink dirs are separated), 
but this time based on Python 2.7.11 x86?

BTW, on https://github.com/streamlink/streamlink/issues/86 you should indicate that:  
1. Your offering won't run on Windows XP, since the last supported 
Python version there is 3.4.3
2. On compatible Windows OSes, 
[Visual C++ Redistributable for Visual Studio 2015](https://www.microsoft.com/en-us/download/details.aspx?id=48145) 
must be installed on the system, if it isn't already!

Thanks in advance!

**Comment by RosadinTV** [comment]
@Vangelis66 
Here is (but of course may have some bugs too): https://1drv.ms/u/s!AteNlrENG5PD9WaVaeDDq1EGj60H
Also updated the original post with the requirements too 😉

**Comment by beardypig** [comment]
I can look it to this if no one else already is.

**Comment by gravyboat** [comment]
@beardypig I don't think anyone is. Seems like most of the 'main' people on the project are very busy this week, myself included.

**Comment by beardypig** [comment]
I took a look, it seems to be a difference in integer division between py2 and py3. I have created a PR with a fix.

**Comment by gravyboat** [comment]
Goign to close since this since #96 resolved this.

**Comment by Vangelis66** [comment]
@RosadinTV 

Many thanks for your Python 2.7.11 (x86) "portable" Windows package; you are being awfully kind! 
This can be run even on Windows XP, since, much like livestreamer-win32, it only requests that MS Visual C++ 2008 Redistributable Package is installed (the case already on most systems!). I couldn't help noticing, however, the significantly larger size of folder "Python 2.7.11" compared to the size of  "Python 3.5.2"  folder in your (second) original release; maybe "Python 2.7.11" could be trimmed further down by omitting unneeded modules (apologies if this is a stupid question, I know nothing about Python)? 

In any case, as you have said, using the Python2 based package does not generate the bug I reported. 
I think the devs do not plan on supporting Python2, only Python3; is there an easy way for me to upgrade your Python3 portable package (posted in https://github.com/streamlink/streamlink/issues/86) so as to include the fix commited by @beardypig and merged by @gravyboat? 
In other words, can the contents of folder "Streamlink" be modified/updated manually without breaking the package as a whole? 

Thanks again!

**Comment by RosadinTV** [comment]
@Vangelis66
You have to patch some files and move specific directories, but later maybe i release a program that makes all the work.
I mean a program that download the whole project with the latest commit and automatically create a ready-to-use release.
If i develop this i will be posting here.

**Comment by gravyboat** [comment]
@Vangelis66 @RosadinTV I wouldn't worry about it too much at this point. Doing a release is pretty easy for us now that everything is set up, we just need to automate the builds a bit more. I think there are a lot more improvements within Streamlink. As of now the Windows installer has everything from master up until today: https://github.com/streamlink/streamlink/releases

**Comment by RosadinTV** [comment]
@gravyboat 
Cool, i'll check this 😄 
Probably i will still develop the automatized build generator for my Project (a GUI that uses ytdl and livestreamer .. but now im moving to streamlink)
PS: This release requires Visual C++ too?
Thanks.

**Comment by gravyboat** [comment]
@RosadinTV Sounds good regarding your tool! Check out https://github.com/bastimeyer/livestreamer-twitch-gui for some of the Twitch stuff. The main dev is a nice guy who might be interested in extending the functionality.

**Comment by RosadinTV** [comment]
@gravyboat Cool, this can be useful, thanks.

**Comment by Vangelis66** [comment]
@RosadinTV wrote: 

> You have to patch some files

... to apply the fix for https://github.com/streamlink/streamlink/issues/89 on your Py3 package, I ended up patching locally 
(with a text editor) files "itvplayer.py" & "livestream.py" according to https://github.com/streamlink/streamlink/commit/c9babf344771da8b49d1b53e5e0b058a8dfe5a50

 > You have to patch some files and move specific directories 
> (snip) 
> a program that downloads the whole project with the latest commit
> and automatically create a ready-to-use release.

Well, I messed around with that idea, this is what I did: 
1. Backed up directory "Streamlink" first. 
2. I fetched [streamlink-master.zip](https://github.com/streamlink/streamlink/archive/master.zip) and unzipped it.
3. I used folders and files from within "streamlink-master" directory to overwrite 
existing folders and files inside original "streamlink" directory, making sure 
the directory structure/hierarchy is kept intact.
4. After a bit of trial-and-error, I discovered that for RTMP plugins (like itvplayer) 
to function correctly, I had to use file "rtmpdump.py" from your original distribution... 

And a tip for other Windows users: 
Instead of launching the Command Prompt and CD-ing to the directory where the package 
resides (or browsing to the directory in Explorer, keeping SHIFT pressed, right-clicking 
the directory and from the context menu selecting "Open Command Window Here"), 
create inside the directory a windows shortcut file (hidden extension .lnk) named 
"Run_StramlinkPortable", its target field should read: 
`C:\Windows\System32\cmd.exe /k Streamlink.bat` 
and its "Start in:" field should be left empty. DC the shortcut and you should be fine! 
(Tested on VistaSP2 & 7SP1 only!).

@gravyboat wrote: 

> As of now, the Windows installer has everything from master up until today: 
> [streamlink-0.0.2.exe](https://github.com/streamlink/streamlink/releases/download/0.0.2/streamlink-0.0.2.exe)

Many thanks, but that's an NSIS installer; if I have a choice, I prefer to not have things installed, 
especially when they permanently modify my PATH environment variable... 
Even in the case of livestreamer, I was/am only using the "zip" distributions 
(**Zip archive** and **Nightly build**), not the installer. 
Do your future plans include a "zip" Windows release or a self-contained package 
like the one kindly compiled by @RosadinTV? 
The latter is happily deployed on an external USB HDD of mine 
(which can then be used on other PCs), downloads streams (-o switch) 
by default to that location, doesn't touch the registry/PATH 
and the only thing written to the host PC is an innocuous file at: 
`%APPDATA%\streamlink\cli.json`

BTW, since your NSIS setup is based on Python 3.5.2, the same 
system requirements apply as with @RosadinTV's package, i.e. 
1. Streamlink won't run on Windows XP, since the last supported
Python version there is 3.4.3
2. On compatible Windows OSes,
Visual C++ Redistributable for Visual Studio 2015
must be installed on the system, if it isn't already! 
(If you wish to support WinXP, currently at ~10% of ALL Windows users, 
you should downgrade to 3.4.3/2.7.11; and VC++ Redistributable for VS2015 
demands at minimum XP SP3...). 

Best regards.

**Comment by gravyboat** [comment]
I don't know if there are plans to create a self contained package like that right now. Regarding XP support it's not something we've discussed. I'm against spending the time on it as we're already pretty limited. The issue is and always will be time, is it worth the dev time to support an already small user base on a very old distro? I don't know, I just know I would rather spend my time working on other more important things for the project. The joy of open source is that if people want these things they can easily develop them and open a PR to get the support added. As such if you are interested in developing and maintaining these releases I'm sure everyone would be happy to review a PR for it.

**Comment by sbstp** [comment]
Full disclosure here, I joined this project because I'm a fairly heavy livestreamer user and I was pissed that it went unmaintained. So I worked on the things required to make it work again for my needs. I hate Windows with a passion and I would really like to avoid installing development tools on it. I'm only using it for gaming/work and watching streams while gaming/working. So the NSIS installer is exactly what fits my needs. A small & self contained package that just works and can be built from linux.

Now, obviously, there are different needs out there. If you want to build the package as you like, I can support you in a reasonable manner, but I won't do the entire work required to do it. I already spent a day trying to get various distribution tools to work and none have. I'm not sure if this project is a really odd case for distribution, but it's surprisingly hard to deal with. I couldn't even get the old system to work because cx_freeze/bbfreeze cannot be installed anymore.

If @RosadinTV wants to make their process reproducable/scriptable, or someone else wants to contribute a distribution script, I'll happily review it and test it.

**Comment by gyvess** [comment]
If it puts any of your minds at ease I can confirm that both the official Streamlink Windows release and Rosadin's Python2.7 release work on Windows XP.  I tested both with Youtube and Twitch.

Only problem I've had is that official is run from a .py and Rosadin's is a .bat.  Both of which act slightly funky when used with the 'Open Livestreamer' Firefox Addon and Livestreamer-Twitch-GUI.  However, I assume that's a matter of adapting the two separate interfaces than something wrong with the builds themselves.
